### ISSUE TITLE
Allow placeholders in matchers

### ISSUE THREAD

**Issue by pascalgn** [caddyserver/caddy#2802]
## 1. What would you like to have changed?

We would like to use placeholders (especially env. placeholders) in the `match` part:

```json
          "routes": [
            {
              "match": [
                {
                  "host": ["{env.DOMAIN_NAME}"]
                }
              ]
```

## 2. Why is this feature a useful, necessary, and/or important addition to this project?

Especially when running in Docker, it can be useful to have an image with the configuration and run it with a different domain, like `docker run -e DOMAIN_NAME=example.com ...`

## 3. What alternatives are there, or what are you doing in the meantime to work around the lack of this feature?

Currently we use some script inside the Docker image which replaces placeholders in the configuration before starting Caddy

## 4. Please link to any relevant issues, pull requests, or other discussions.

None that I know of

**Comment by pascalgn** [comment]
This is for v2, by the way!

I would give it a shot, but I'm not sure where to add the ReplaceAll call...

**Comment by mholt** [comment]
Sure, I can see why this would be useful.

The replacement would probably happen in between these lines: https://sourcegraph.com/github.com/caddyserver/caddy@fe36d26b63b6398592e46604d1795f84ce0477d4/-/blob/modules/caddyhttp/matchers.go#L122-123

**Comment by pascalgn** [comment]
I had a look, but I'm not sure about this: When adding it in matchers.go, it would mean that every incoming request would get a (maybe small?) performance hit due to the replacement logic.

I was thinking more about adding it in the configuration setup, maybe in https://github.com/caddyserver/caddy/blob/v2/modules/caddyhttp/routes.go#L178? Or maybe at the end of loadConfig, https://github.com/caddyserver/caddy/blob/v2/cmd/main.go#L96? WDYT?

**Comment by mholt** [comment]
@pascalgn Replacers in v2 are very efficient: https://github.com/caddyserver/caddy/pull/2674

> (maybe small?) performance hit

You can test this. [Write a benchmark](https://golang.org/pkg/testing/#hdr-Benchmarks) for the Host matcher without replacer support, then add the ReplaceAll call and run the benchmark again. Post the results here.

**Comment by pascalgn** [comment]
Getting the numbers is easy, but making sense of them... 😄

Without placeholder replacements (status quo)

```
$ go test -bench .
goos: darwin
goarch: amd64
pkg: github.com/caddyserver/caddy/v2/modules/caddyhttp
BenchmarkHostMatcherWithoutPlaceholders-8   	15595172	        74.6 ns/op
PASS
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp	1.261s
```

With placeholder replacement (after my change)

```
$ go test -bench .
goos: darwin
goarch: amd64
pkg: github.com/caddyserver/caddy/v2/modules/caddyhttp
BenchmarkHostMatcherWithoutPlaceholders-8   	 5592524	       216 ns/op
PASS
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp	1.446s
```

(tests run on 8-core Macbook Pro)

**Comment by mholt** [comment]
What is your code? The benchmark is only useful if we know what it is measuring.

**Comment by pascalgn** [comment]
Code changes are here: https://github.com/caddyserver/caddy/pull/2810/files

**Comment by mholt** [comment]
(I was mobile earlier, so now that I'm at my computer, I can write more)

I wanted to answer this:

> I was thinking more about adding it in the configuration setup, maybe in https://github.com/caddyserver/caddy/blob/v2/modules/caddyhttp/routes.go#L178? Or maybe at the end of loadConfig, https://github.com/caddyserver/caddy/blob/v2/cmd/main.go#L96? WDYT?

The reason this isn't going to work is because replacers are dependent on context. So if you wanted to perform a replacement based on something in the HTTP request, you can't do that if the replacement happens before any HTTP requests come in. So it has to be during the HTTP request.

**Comment by pascalgn** [comment]
One more test with placeholder replacement (this time actually replacing something)

```
$ go test -bench .
goos: darwin
goarch: amd64
pkg: github.com/caddyserver/caddy/v2/modules/caddyhttp
BenchmarkHostMatcherWithoutPlaceholders-8   	 5529973	       215 ns/op
BenchmarkHostMatcherWithPlaceholders-8      	 2839238	       414 ns/op
PASS
ok  	github.com/caddyserver/caddy/v2/modules/caddyhttp	3.039s
```

**Comment by mholt** [comment]
Great, thanks for that test.

It actually performs faster when you obtain the replacer from the http.Request, than when you make a new one each time.

I ran the latest benchmark locally and established a baseline on my system.

Here is the baseline (no changes):

```
BenchmarkHostMatcherBaseline-8   	13842432	        84.3 ns/op	      32 B/op	       1 allocs/op
```

With your proposed changes, including calling `NewReplacer()` in the matcher (incorrect, but included here for comparison):

```
BenchmarkHostMatcherWithNewReplacer-8   	 5137562	       227 ns/op	     144 B/op	       5 allocs/op
```

Now when `NewReplacer()` (incorrect) is changed to `r.Context().Value(caddy.ReplacerCtxKey).(caddy.Replacer)` (correct):

```
BenchmarkHostMatcherWithExistingReplacer-8   	 9837710	       118 ns/op	      32 B/op	       1 allocs/op
```

Performance is already significantly better. All of these had "localhost" as the matcher string, so let's try one with an env var placeholder, like your proposed changes:

```
BenchmarkHostMatcherWithPlaceholders-8   	 4093802	       287 ns/op	      64 B/op	       2 allocs/op
```

As expected, a slight perf hit but still on the order of tens of nanoseconds.

In any case, I think this goes to show that replacers and placeholders are very, very efficient. When we're striving to shave off nanoseconds... well... that's already pretty good. If these were an order of magnitude different, I'd be concerned, but this is nothing to be concerned about with _network software_ where the primary bottleneck is usually socket I/O.

Anyway, this feature is "nice to have" but maybe until someone needs it, we should hold out on it. It's mostly a simple change, but as you noted, it adds one more nuance to automatic HTTPS. We can implement it later if someone needs it, sound good?

**Comment by pascalgn** [comment]
I still think the feature is quite useful, I just think it's not useful to include HTTP placeholders. I had a look again and a HostMatch with e.g. `{env.DOMAIN}` or even `{system.hostname}` seems very useful.

On the other hand, having `{http.request.method}.example.com` or `{http.request.scheme}.example.com` or anything else from the `http.` placeholders seems exotic at best.

So I think we should implement this feature with only the placeholders from `globalDefaultReplacements` for now and this should already cover a lot of use cases, IMO. Performance still needs to be considered, to minimize the amount of allocations and general impact per request, of course.

WDYT?

**Comment by mholt** [comment]
@pascalgn Okay, I do see the utility in that.

If the replacements were going to be implemented out-of-band of HTTP requests, it would be done in the `MatchHost` module's `Provision()` function (see lifecycle of modules, here: https://github.com/caddyserver/caddy/wiki/v2:-Writing-a-Module#module-lifecycle).

However, we're talking a few tens of nanoseconds, so the performance hit is negligible. If we have a customer/user who says otherwise (and we can show that it's in the Host matcher and not somewhere else), then we can consider alternatives. But for now, let's make this consistent with some of the other matchers which accept placeholders, and they are evaluated at request-time.

If you happen to be available to make the change in the next hour or two, I can merge it in and get it out with beta 6 which I'm tagging today. If not, I'll push a commit to your branch and merge it in.

Thanks!

**Comment by pascalgn** [comment]
Sorry, just read this! If you're not already working on it, I would give it a go. I hope to have some working code in ~30 minutes! 👍

**Comment by mholt** [comment]
@pascalgn Go for it! I'll be standing by to review, or if you have questions :)

**Comment by pascalgn** [comment]
@mholt Done, see #2810

**Comment by pascalgn** [comment]
@mholt This is done now and I think it's fine! However, I think there's one more point we could improve (not necessarily urgent, IMO). When the user forgets to set the environment variable, no replacement happens, which leads to error messages like this: `2019/10/14 17:02:45 [ERROR][{env.DOMAIN_NAME}] failed to obtain certificate: acme: error`. This is because we only have `ReplaceAll`, which doesn't return an error when the replacement is not found. So I think we could maybe have a variant which allows us to detect when a replacement was unsuccessful and then fail during Caddy startup, or at least in `automaticHTTPS` when the domains are set up. WDYT?

**Comment by mholt** [comment]
@pascalgn Hm, yeah... I've just pushed a commit to CertMagic to reject hosts with certain special characters that are commonly found in configuration, but which are not allowed in domain names: https://github.com/mholt/certmagic/commit/96223721b8dd7ddb8eb79a4824541f711cf01d4f

That should at least skip host values which are not properly configured.

I do agree though that ignoring unrecognized placeholders (i.e. producing no error) is not always ideal (but sometimes is). We can change ReplaceAll or the Replacer logic to fix this. Want to open a new issue to propose a change? (What do you think we should change in order to emit errors?)
